### PR TITLE
Support for returning array-like types from the Execute(Async)Script wd command

### PR DESCRIPTION
**Pull request by georgeroman**
---

- [X] `./mach build -d` does not report any errors
- [X] `./mach test-tidy` does not report any errors

- [X] There are tests for these changes

---
This change is [Reviewable](https://reviewable.io/reviews/servo/servo/23725)

### CODE REVIEW

File: components/script/webdriver_handlers.rs [hunk 1]
```diff
@@ -81,8 +84,42 @@ pub unsafe fn jsval_to_webdriver(cx: *mut JSContext, val: HandleValue) -> WebDri
...
+            let mut result: Vec<WebDriverJSValue> = Vec::new();
+
+            let length =
+                match get_property::<u32>(cx, object_handle, "length", ConversionBehavior::Default)
+                {
+                    Ok(length) => match length {
+                        Some(length) => length,
+                        _ => return Err(WebDriverJSError::UnknownType),
+                    },
+                    _ => return Err(WebDriverJSError::UnknownType),
```
Comment (jdm) [inline components/script/webdriver_handlers.rs]: In the Err case we should call throw_dom_exception.

File: components/script/webdriver_handlers.rs [hunk 2]
```diff
@@ -81,8 +84,42 @@ pub unsafe fn jsval_to_webdriver(cx: *mut JSContext, val: HandleValue) -> WebDri
...
+                };
+
+            for i in 0..length {
+                rooted!(in(cx) let mut item = UndefinedValue());
+                match get_property_jsval(cx, object_handle, &i.to_string(), item.handle_mut()) {
+                    Ok(_) => match jsval_to_webdriver(cx, item.handle()) {
+                        Ok(converted_item) => result.push(converted_item),
+                        _ => return Err(WebDriverJSError::UnknownType),
+                    },
+                    _ => return Err(WebDriverJSError::UnknownType),
```
Comment (jdm) [inline components/script/webdriver_handlers.rs]: In the Err case we should call throw_dom_exception.

File: components/script/webdriver_handlers.rs [hunk 3]
```diff
@@ -81,8 +84,42 @@ pub unsafe fn jsval_to_webdriver(cx: *mut JSContext, val: HandleValue) -> WebDri
...
-        Ok(WebDriverJSValue::Null)
+    } else if val.get().is_object() {
+        let object: *mut JSObject = match FromJSValConvertible::from_jsval(cx, val, ()).unwrap() {
+            ConversionResult::Success(object) => object,
+            _ => unreachable!(),
+        };
+        let _ac = JSAutoRealm::new(cx, object);
+        let object_handle = Handle::new(&object);
+
+        if is_array_like(cx, val) {
```
Comment (jdm) [inline components/script/webdriver_handlers.rs]: Let's do this instead:
```rust
if !is_array_like(cx, val) {
    return Err(WebDriverJSError::UnknownType);
}
```

File: components/script/webdriver_handlers.rs [hunk 4]
```diff
@@ -81,8 +84,42 @@ pub unsafe fn jsval_to_webdriver(cx: *mut JSContext, val: HandleValue) -> WebDri
                 _ => unreachable!(),
             };
         Ok(WebDriverJSValue::String(String::from(string)))
-    } else if val.get().is_null() {
-        Ok(WebDriverJSValue::Null)
+    } else if val.get().is_object() {
+        let object: *mut JSObject = match FromJSValConvertible::from_jsval(cx, val, ()).unwrap() {
```
Comment (jdm) [inline components/script/webdriver_handlers.rs]: Let's use `rooted!(in(cx) let object = ...);` here, and then we can use `object.handle()` later instead.

**Review by jdm — CHANGES_REQUESTED** [review]
Great work here! We can write the code to be a bit more safe, though, particularly when there are problems converting values.

File: components/script/webdriver_handlers.rs [hunk 5]
```diff
@@ -81,8 +89,46 @@ pub unsafe fn jsval_to_webdriver(cx: *mut JSContext, val: HandleValue) -> WebDri
...
+            return Err(WebDriverJSError::UnknownType);
+        }
+
+        let mut result: Vec<WebDriverJSValue> = Vec::new();
+
+        let length =
+            match get_property::<u32>(cx, object.handle(), "length", ConversionBehavior::Default) {
+                Ok(length) => match length {
+                    Some(length) => length,
+                    Err(error) => {
```
Comment (jdm) [inline components/script/webdriver_handlers.rs]: I do not believe there is an enum with `Some/Err` variants ;)

File: components/script/webdriver_handlers.rs [hunk 6]
```diff
@@ -81,8 +89,46 @@ pub unsafe fn jsval_to_webdriver(cx: *mut JSContext, val: HandleValue) -> WebDri
...
+                        throw_dom_exception(cx, global_scope, error);
```
Comment (jdm) [inline components/script/webdriver_handlers.rs]: I'm pretty sure this belongs in the outer Err block instead.

File: components/script/webdriver_handlers.rs [hunk 7]
```diff
@@ -81,8 +89,46 @@ pub unsafe fn jsval_to_webdriver(cx: *mut JSContext, val: HandleValue) -> WebDri
...
+                _ => return Err(WebDriverJSError::UnknownType),
+            };
+
+        for i in 0..length {
+            rooted!(in(cx) let mut item = UndefinedValue());
+            match get_property_jsval(cx, object.handle(), &i.to_string(), item.handle_mut()) {
+                Ok(_) => match jsval_to_webdriver(cx, global_scope, item.handle()) {
+                    Ok(converted_item) => result.push(converted_item),
+                    Err(error) => {
+                        throw_dom_exception(cx, global_scope, error);
```
Comment (jdm) [inline components/script/webdriver_handlers.rs]: I believe this belongs in the outer Err block instead.

**Review by jdm — CHANGES_REQUESTED** [review]

File: components/script/webdriver_handlers.rs (hunk 5 above)
Comment (georgeroman) [reply components/script/webdriver_handlers.rs]: Ups, looks like I was too sleepy when I wrote that :D

**Comment by jdm** [comment]
```
{
    "status": "PASS", 
    "group": "default", 
    "message": null, 
    "stack": null, 
    "subtest": "test_format_and_dimensions", 
    "test": "/webdriver/tests/take_screenshot/screenshot.py", 
    "line": 14752, 
    "action": "test_result", 
    "expected": "FAIL"
}
```
This passed on both mac and linux.

**Comment by jdm** [comment]
Also `./mach fmt` has some changes for you:
```
Diff in /repo/components/script/webdriver_handlers.rs at line 152:
                 Ok(length) => match length {
                     Some(length) => length,
                     _ => return Err(WebDriverJSError::UnknownType),
-                }
(B+                },
(B                 Err(error) => {
                     throw_dom_exception(cx, global_scope, error);
                     return Err(WebDriverJSError::UnknownType);
Diff in /repo/components/script/webdriver_handlers.rs at line 165:
                 Ok(_) => match jsval_to_webdriver(cx, global_scope, item.handle()) {
                     Ok(converted_item) => result.push(converted_item),
                     err @ Err(_) => return err,
-                }
(B+                },
(B                 Err(error) => {
                     throw_dom_exception(cx, global_scope, error);
                     return Err(WebDriverJSError::UnknownType);
```

**Comment by jdm** [comment]
This PR will need to be rebased.

File: tests/wpt/metadata/webdriver/tests/execute_script/collections.py.ini [hunk 8]
```diff
@@ -14,9 +14,6 @@
   [test_html_collection]
     expected: FAIL
```
Comment (jdm) [inline tests/wpt/metadata/webdriver/tests/execute_script/collections.py.ini]: It looks like array-like objects that are not arrays aren't working yet?

Comment (georgeroman) [reply tests/wpt/metadata/webdriver/tests/execute_script/collections.py.ini]: Thanks for noticing that! Yeah, it seems that I forgot to handle the case where the array's elements are WebElements.

**Comment by georgeroman** [comment]
https://build.servo.org/builders/linux-rel-css/builds/12988/steps/shell__4/logs/filtered-wpt-errorsummary.log
This seems unrelated.

@bors-servo try=wpt

**Comment by georgeroman** [comment]
@jdm Any needed changes for this?

**Comment by CYBAI** [comment]
```
  ▶ Unexpected subtest result in /webdriver/tests/execute_script/cyclic.py:
  │ FAIL [expected PASS] test_element_in_collection
  │   → AssertionError: unsupported operation (500): Unsupported return type
  │ 
  │ session = <Session 3f7c05ab-79c8-4cf0-86cc-5cfbab428be7>
  │ 
  │     def test_element_in_collection(session):
  │         session.url = inline("<div></div>")
  │         divs = session.find.css("div")
  │     
  │         response = execute_script(session, """
  │             let div = document.querySelector("div");
  │             div.reference = div;
  │             return [div];
  │             """)
  │ >       value = assert_success(response)
  │ 
  │ divs       = [<Element 148e9477fe0a4addb5d5e53822bd4e68>]
  │ response   = <Response status=500 error=<UnsupportedOperationException http_status=500>>
  │ session    = <Session 3f7c05ab-79c8-4cf0-86cc-5cfbab428be7>
  │ 
  │ tests/wpt/web-platform-tests/webdriver/tests/execute_script/cyclic.py:61: 
  │ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
  │ 
  │ response = <Response status=500 error=<UnsupportedOperationException http_status=500>>
  │ value = None
  │ 
  │     def assert_success(response, value=None):
  │         """
  │         Verify that the provided webdriver.Response instance described
  │         a valid success response as defined by `dfn-send-a-response` and
  │         the provided response value.
  │     
  │         :param response: ``webdriver.Response`` instance.
  │         :param value: Expected value of the response body, if any.
  │         """
  │ >       assert response.status == 200, str(response.error)
  │ E       AssertionError: unsupported operation (500): Unsupported return type
  │ E         
  │ E       assert 500 == 200
  │ E        +  where 500 = <Response status=500 error=<UnsupportedOperationException http_status=500>>.status
  │ 
  │ response   = <Response status=500 error=<UnsupportedOperationException http_status=500>>
  │ value      = None
  │ 
  └ tests/wpt/web-platform-tests/webdriver/tests/support/asserts.py:67: AssertionError
```

**Comment by georgeroman** [comment]
This is strange, the test passes locally and I don't see a reason for it to fail on the CI.

**Comment by CYBAI** [comment]
👀

```
  ▶ Unexpected subtest result in /webdriver/tests/execute_script/cyclic.py:
  │ FAIL [expected PASS] test_element_in_collection
  │   → AssertionError: unsupported operation (500): Unsupported return type
  │ 
  │ session = <Session 368972d9-5331-42dd-bb13-d931cb0b327e>
  │ 
  │     def test_element_in_collection(session):
  │         session.url = inline("<div></div>")
  │         divs = session.find.css("div")
  │     
  │         response = execute_script(session, """
  │             let div = document.querySelector("div");
  │             div.reference = div;
  │             return [div];
  │             """)
  │ >       value = assert_success(response)
  │ 
  │ divs       = [<Element ec87c363e4694d61b3590c0dbd8e44b6>]
  │ response   = <Response status=500 error=<UnsupportedOperationException http_status=500>>
  │ session    = <Session 368972d9-5331-42dd-bb13-d931cb0b327e>
  │ 
  │ tests/wpt/web-platform-tests/webdriver/tests/execute_script/cyclic.py:61: 
  │ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
  │ 
  │ response = <Response status=500 error=<UnsupportedOperationException http_status=500>>
  │ value = None
  │ 
  │     def assert_success(response, value=None):
  │         """
  │         Verify that the provided webdriver.Response instance described
  │         a valid success response as defined by `dfn-send-a-response` and
  │         the provided response value.
  │     
  │         :param response: ``webdriver.Response`` instance.
  │         :param value: Expected value of the response body, if any.
  │         """
  │ >       assert response.status == 200, str(response.error)
  │ E       AssertionError: unsupported operation (500): Unsupported return type
  │ E         
  │ E       assert 500 == 200
  │ E        +  where 500 = <Response status=500 error=<UnsupportedOperationException http_status=500>>.status
  │ 
  │ response   = <Response status=500 error=<UnsupportedOperationException http_status=500>>
  │ value      = None
  │ 
  └ tests/wpt/web-platform-tests/webdriver/tests/support/asserts.py:67: AssertionError
```

**Comment by georgeroman** [comment]
Hmm, this is really weird. The test passes in https://build.servo.org/builders/linux-rel-wpt/builds/13078 but fails on [taskcluster](https://tools.taskcluster.net/groups/SnojOy35TmqnGGn5YY4G_Q). Is there any notable difference between the two that might be the cause of this behavior?

**Comment by jdm** [comment]
The taskclyster tests really n in macOS, but it's not clear to me why that would matter. I can try running them on my local macOS setup to check.

**Comment by CYBAI** [comment]
@bors-servo try=wpt-mac

- Let's try it?

**Comment by jdm** [comment]
@retry

**Comment by jdm** [comment]
I have not been able to reproduce the taskcluster failure yet. I do suggest making this change in order to pass more tests, however:
```
diff --git a/components/script/webdriver_handlers.rs b/components/script/webdriver_handlers.rs
index cf103c90a3..5e369c19bc 100644
--- a/components/script/webdriver_handlers.rs
+++ b/components/script/webdriver_handlers.rs
@@ -165,7 +165,7 @@ pub unsafe fn jsval_to_webdriver(
                 },
                 Err(error) => {
                     throw_dom_exception(cx, global_scope, error);
-                    return Err(WebDriverJSError::UnknownType);
+                    return Err(WebDriverJSError::JSError);
                 },
             };
 
@@ -178,7 +178,7 @@ pub unsafe fn jsval_to_webdriver(
                 },
                 Err(error) => {
                     throw_dom_exception(cx, global_scope, error);
-                    return Err(WebDriverJSError::UnknownType);
+                    return Err(WebDriverJSError::JSError);
                 },
             }
         }
diff --git a/components/script_traits/webdriver_msg.rs b/components/script_traits/webdriver_msg.rs
index 586ea4a444..c33a5ae302 100644
--- a/components/script_traits/webdriver_msg.rs
+++ b/components/script_traits/webdriver_msg.rs
@@ -76,6 +76,7 @@ pub enum WebDriverJSValue {
 pub enum WebDriverJSError {
     Timeout,
     UnknownType,
+    JSError,
     /// Occurs when handler received an event message for a layout channel that is not
     /// associated with the current script thread
     BrowsingContextNotFound,
diff --git a/components/webdriver_server/lib.rs b/components/webdriver_server/lib.rs
index e67d9ca746..fce4edff59 100644
--- a/components/webdriver_server/lib.rs
+++ b/components/webdriver_server/lib.rs
@@ -1401,6 +1401,10 @@ impl Handler {
                 ErrorStatus::UnsupportedOperation,
                 "Unsupported return type",
             )),
+            Err(WebDriverJSError::JSError) => Err(WebDriverError::new(
+                ErrorStatus::JavascriptError,
+                "JS evaluation raised an exception",
+            )),
             Err(WebDriverJSError::BrowsingContextNotFound) => Err(WebDriverError::new(
                 ErrorStatus::JavascriptError,
                 "Pipeline id not found in browsing context",
```

**Comment by jdm** [comment]
A release build of this branch did not reproduce the taskcluster failure either. Peculiar!

**Comment by georgeroman** [comment]
I added the changes you suggested. Maybe with these in place we'll get a `JavascriptError` so that we can be sure that the failure is triggered by a JS error.

**Comment by jdm** [comment]
It is indeed hitting a JSError response. Interesting. I tried reproducing on one of the CI machines and still couldn't.

**Comment by georgeroman** [comment]
@jdm Maybe try this one more time?

**Comment by CYBAI** [comment]
I saw the JS error 👀

```
{
  "status": "FAIL",
  "group": "default",
  "message": "AssertionError: javascript error (500): JS evaluation raised an exception",
  "stack": "session = <Session 22d896b6-493c-40f3-a4c8-7974daf9afca>\n\n    def test_element_in_collection(session):\n        session.url = inline(\"<div></div>\")\n        divs = session.find.css(\"div\")\n    \n        response = execute_script(session, \"\"\"\n            let div = document.querySelector(\"div\");\n            div.reference = div;\n            return [div];\n            \"\"\")\n>       value = assert_success(response)\n\ndivs       = [<Element 607d0458df0f450eb835d66c731d9f4d>]\nresponse   = <Response status=500 error=<JavascriptErrorException http_status=500>>\nsession    = <Session 22d896b6-493c-40f3-a4c8-7974daf9afca>\n\ntests/wpt/web-platform-tests/webdriver/tests/execute_script/cyclic.py:61: \n_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ \n\nresponse = <Response status=500 error=<JavascriptErrorException http_status=500>>\nvalue = None\n\n    def assert_success(response, value=None):\n        \"\"\"\n        Verify that the provided webdriver.Response instance described\n        a valid success response as defined by `dfn-send-a-response` and\n        the provided response value.\n    \n        :param response: ``webdriver.Response`` instance.\n        :param value: Expected value of the response body, if any.\n        \"\"\"\n>       assert response.status == 200, str(response.error)\nE       AssertionError: javascript error (500): JS evaluation raised an exception\nE         \nE       assert 500 == 200\nE        +  where 500 = <Response status=500 error=<JavascriptErrorException http_status=500>>.status\n\nresponse   = <Response status=500 error=<JavascriptErrorException http_status=500>>\nvalue      = None\n\ntests/wpt/web-platform-tests/webdriver/tests/support/asserts.py:67: AssertionError",
  "subtest": "test_element_in_collection",
  "test": "/webdriver/tests/execute_script/cyclic.py",
  "line": 7725,
  "action": "test_result",
  "expected": "PASS"
}
```

**Comment by georgeroman** [comment]
Ok, I added an issue for it. This is holding this PR and the next one for the rest of the functionality back, so I suggest to let it go in this particular PR. What do you think? Is there a way to ignore only this specific failure without ignoring all tests in `cyclic.py`?

**Comment by CYBAI** [comment]
> Is there a way to ignore only this specific failure without ignoring all tests in `cyclic.py` ?

https://github.com/servo/servo/blob/555fa75b2c7f4faa0818e86296a4ea3d5b8e4d42/tests/wpt/mozilla/meta/css/pixel_snapping_glyphs.html.ini#L3-L4

**Comment by jdm** [comment]
I agree; it's mysterious but solving it shouldn't be our top priority.

**Comment by georgeroman** [comment]
Marked the test as failing on Mac. This should now be finally ready.

**Comment by jdm** [comment]
```
error[E0308]: mismatched types
   --> components/script/webdriver_handlers.rs:167:41
    |
167 |                     throw_dom_exception(cx, global_scope, error);
    |                                         ^^ expected struct `script_runtime::JSContext`, found *-ptr
    |
    = note: expected type `script_runtime::JSContext`
               found type `*mut js::jsapi::JSContext`
error[E0308]: mismatched types
   --> components/script/webdriver_handlers.rs:180:41
    |
180 |                     throw_dom_exception(cx, global_scope, error);
    |                                         ^^ expected struct `script_runtime::JSContext`, found *-ptr
    |
    = note: expected type `script_runtime::JSContext`
               found type `*mut js::jsapi::JSContext`
error: aborting due to 2 previous errors
```
This needs to be rebased over https://github.com/servo/servo/pull/23872.

**Comment by georgeroman** [comment]
Done.

**Comment by jdm** [comment]
This needs to be updated to accommodate #23941, sadly:
```
diff --git a/Cargo.lock b/Cargo.lock
index 01766ccf8..2d9696423 100644
--- a/Cargo.lock
+++ b/Cargo.lock
@@ -5521,6 +5521,27 @@ dependencies = [
  "lazy_static 1.3.0 (registry+https://github.com/rust-lang/crates.io-index)",
 ]
 
+[[package]]
+name = "webdriver"
+version = "0.39.0"
+source = "registry+https://github.com/rust-lang/crates.io-index"
+dependencies = [
+ "base64 0.10.1 (registry+https://github.com/rust-lang/crates.io-index)",
+ "cookie 0.11.0 (registry+https://github.com/rust-lang/crates.io-index)",
+ "futures 0.1.23 (registry+https://github.com/rust-lang/crates.io-index)",
+ "http 0.1.17 (registry+https://github.com/rust-lang/crates.io-index)",
+ "hyper 0.12.19 (registry+https://github.com/rust-lang/crates.io-index)",
+ "log 0.4.6 (registry+https://github.com/rust-lang/crates.io-index)",
+ "regex 1.1.6 (registry+https://github.com/rust-lang/crates.io-index)",
+ "serde 1.0.88 (registry+https://github.com/rust-lang/crates.io-index)",
+ "serde_derive 1.0.88 (registry+https://github.com/rust-lang/crates.io-index)",
+ "serde_json 1.0.13 (registry+https://github.com/rust-lang/crates.io-index)",
+ "time 0.1.42 (registry+https://github.com/rust-lang/crates.io-index)",
+ "tokio 0.1.11 (registry+https://github.com/rust-lang/crates.io-index)",
+ "unicode-segmentation 1.2.0 (registry+https://github.com/rust-lang/crates.io-index)",
+ "url 1.7.2 (registry+https://github.com/rust-lang/crates.io-index)",
+]
+
 [[package]]
 name = "webdriver"
 version = "0.40.0"
@@ -6337,6 +6358,7 @@ dependencies = [
 "checksum wayland-protocols 0.21.4 (registry+https://github.com/rust-lang/crates.io-index)" = "d3f6cebb98963f028d397e9bad2acf9d3b2f6b76fae65aea191edd9e7c0df88c"
 "checksum wayland-scanner 0.21.4 (registry+https://github.com/rust-lang/crates.io-index)" = "f1927ee62e4e149c010dc9eca8ca47e238416cd6f45f688eb9f8a8e9c3794c30"
 "checksum wayland-sys 0.21.4 (registry+https://github.com/rust-lang/crates.io-index)" = "ca41ed78a12256f81df6f53fcbe4503213ba442e02cdad3c9c888a64a668eaf4"
+"checksum webdriver 0.39.0 (registry+https://github.com/rust-lang/crates.io-index)" = "0533b0b0a05e2e5c081317759a038482806c6085c9605dded03c8bbd2498b042"
 "checksum webdriver 0.40.0 (registry+https://github.com/rust-lang/crates.io-index)" = "6949bd7190fdd0eeeaf24d6103fec026856e60d886b2e64f427da1a0f5e590d3"
 "checksum webrender 0.60.0 (git+https://github.com/servo/webrender)" = "<none>"
 "checksum webrender_api 0.60.0 (git+https://github.com/servo/webrender)" = "<none>"
```

**Pull request merged**